geniza/corpus/models.py

```diff
@@ -925,6 +925,65 @@ def title(self):
 
+    def dating_range(self):
+        """
+        Return the start and end of the document's possible date range, as PartialDate objects,
+        including standardized document dates and inferred Datings, if any exist.
+        """
+        # it is unlikely, but technically possible, that a document could have both on-document
+        # dates and inferred datings, so find the min and max out of all of them.
+        start_date = self.start_date if self.start_date else None
+        end_date = self.end_date if self.end_date else None
+        dating_range = [start_date, end_date]
+
+        # bail out if we don't have any inferred datings
+        if not self.dating_set.exists():
+            return tuple(dating_range)
+
+        # loop through inferred datings to find min and max
+        for dating in self.dating_set.all():
+            # get start from standardized date range (formatted as "date1/date2" or "date")
+            split_date = dating.standard_date.split("/")
+            start = PartialDate(split_date[0])
+            # use numeric format to compare to current min, replace if smaller
+            start_numeric = int(start.numeric_format(mode="min"))
+            if dating_range[0] is None or start_numeric < int(
+                dating_range[0].numeric_format(mode="min")
+            ):
+                dating_range[0] = start
+            # get end from standardized date range
+            end = PartialDate(split_date[1]) if len(split_date) > 1 else start
+            # use numeric format to compare to current max, replace if larger
+            end_numeric = int(end.numeric_format(mode="max"))
+            if dating_range[1] is None or end_numeric > int(
+                dating_range[1].numeric_format(mode="max")
+            ):
+                dating_range[1] = end
+
+        return tuple(dating_range)
+
+    def solr_dating_range(self):
+        """Return the document's dating range, including inferred, as a Solr date range."""
+        dating_range = self.dating_range()
+        solr_dating_range = []
+        if dating_range[0] is not None:
+            # min from dating_range, formatted YYYY-MM-DD or YYYY-MM or YYYY
+            solr_dating_range.append(dating_range[0].isoformat())
+        # if there's a max date in the range, ensure it's not the same as the min
+        if (
+            dating_range[1] is not None
+            and dating_range[1].isoformat(mode="max") != solr_dating_range[0]
+        ):
+            # max from dating_range, formatted YYYY-MM-DD or YYYY-MM or YYYY
+            solr_dating_range.append(dating_range[1].isoformat(mode="max"))
+        if not solr_dating_range:
+            return None
+        # if a single date instead of a range, start and end are the same
+        if len(solr_dating_range) == 1:
+            return solr_dating_range[0]
+        # if there's more than one date, return as a range
+        return "[%s TO %s]" % tuple(solr_dating_range)
```

Comment:
It looks like you support `None` values for start/end in the dating range — I interpreted that as open-ended ranges, but maybe I'm misinterpreting because I don't have full context? Just wondering if instead of omitting the `None` values you should preserve them and convert them to `*` in the range value (as long as at least one value is present).

Reply:
I don't think we support setting open-ended date ranges on documents, inferred or on-document; but we do support setting a single date as opposed to a range. And this is for indexing that backend data, not searching with an open-ended range. 

In case this is a source of confusion: I don't think we'll end up with `None` in min or max as a result of `Document.dating_range`, they're really just in that method as placeholder values for when `Document.start_date` or `Document.end_date` aren't returning something useful. (I actually think those methods just return `None` if not a date… but I wasn't 100% sure about their possible outputs.)

I think the only actual cases we'll encounter are `"start_date/end_date"` or just `"date"`, for both document date and inferred date, so `dating_range` could probably be simplified to reflect that. But essentially it will either return `(start, end)`, if there is some kind of range represented between those two fields, or `(date, date)` with the same date repeated, if there is exactly one date inferred/on-document, and therefore the range of possible dates for that document is exactly that narrow. Then, `solr_dating_range` might get one repeated date `(date, date)` as input, and if those turn out to be the same `date.isoformat(mode="max")` and `mode="min"` then we only want to index that document as having a single date, not an open-ended range. (Though if, for example, the input is just a year repeated `(YYYY, YYYY)`, then it should return a range like `[YYYY-01-01 TO YYYY-12-31]`.)